ops/_main.py

```diff
@@ -217,43 +151,24 @@ def __init__(self, charm_dir: Path, juju_context: _JujuContext):
         self._charm_dir = charm_dir
         self._exec_path = Path(self._juju_context.dispatch_path or sys.argv[0])
 
-        dispatch = charm_dir / 'dispatch'
-        if self._juju_context.version.is_dispatch_aware() and _exe_path(dispatch) is not None:
-            self._init_dispatch()
-        else:
-            self._init_legacy()
+        # Grab the correct hook from JUJU_DISPATCH_PATH, e.g. hooks/install.
+        self._dispatch_path = Path(self._juju_context.dispatch_path)
 
-    def ensure_event_links(self, charm: '_charm.CharmBase'):
-        """Make sure necessary symlinks are present on disk."""
-        if self.is_dispatch_aware:
-            # links aren't needed
-            return
+        if 'OPERATOR_DISPATCH' in os.environ:
+            logger.debug('Charm called itself via %s.', self._dispatch_path)
+            raise _Abort(0)
+        os.environ['OPERATOR_DISPATCH'] = '1'
```

Comment:
Do we have a test for this behaviour? I couldn't find one, but test_main is complicated, so I could have missed it.

Reply:
Arguably, that's not the subject of this PR, whether or not there is a test.